### DISCUSSION TITLE
Can I use soft assertions in behave (like in Java TestNG)

### DISCUSSION THREAD

**Discussion by JSorella** [behave/behave#1094] [Q&A]
Hi,
I want to mark a step as failed but continue with the execution.

For example:

```gherkin
Scenario: Search results validations

    When I login into the website 
    Then I should be in "Search" page
    When I input "test" in the Search input field
    And I click over the search button
    Then I should see that 5 results are shown
    And I should see that the first result frame have no content
```
... then consider this step implementation:

```python
@step('I should see that {expected_quantity} results are shown')
def step_impl(context, expected_quantity):
    actual_results_quantity = len(context.page.get_all_results())
    try:
        assert_that(actual_results_quantity, expected_quantity)
    except AssertionError:
        logger.error(
            f'Expected {condition} results but {actual_frames_quantity} were shown.'
        )
        # JE-HINT: UNCOMMENT THE FOLLOWING LINE: To re raise the AssertionError from above: 
        # raise
```

As you might be thinking, I added the try / catch block  in order to explain you what I need. But, in the `except` block, I want to add an instruction to tell Behave "hey, don't mark this step as passed!! I want to show this step as failed in order to explain this is wrong, but don't stop there, because I need to validate if there is a first result with no content!".

Is this possible?
Is there any way to achieve this "cheating" with the framework?

Thanks!!

**Comment by jenisys** [comment]
SEE ALSO: As example/solution for this use case
* [features/runner.continue_after_failed_step.feature](https://github.com/behave/behave/blob/main/features/runner.continue_after_failed_step.feature)

NOTES:
* In general, it is not advised to do that 
  (due to correlated failures and stricter checking is necessary in step implementations that preconditions/assumptions are met)
* You should still pass the AssertionError exception to the runner (to normally mark the step as failed)
* For development/diagnostics, it is probably a good idea to use a special tag, like `@runner.continue_after_failed_step`, that marks those scenarios that should continue after a failed step(s)

**HINT:**
**Read the detailed discussion below** (for the answer)

**Reply by JSorella** [reply]
Thanks for your time to write this reply. Some notes:

1) I want to work as in Java's TestNG, using soft & hard asserts (mixed in one scenario). This is a commonly used practice.
2) The `runner.continue_after_failed_step` never worked for me. Sill breaks the execution of the scenario (may I need to configure something more?).
3) ChatGPT recommended me this, but didn't work for me:

![image](https://user-images.githubusercontent.com/2180155/227271294-79a8ca76-9db8-4379-8c4f-d93d65cc2a27.png)

**Reply by jenisys** [reply]
Mmh, interesting.
Sorry for my misunderstanding of your initial question.
Actually, there are two different cases/functionalities:

* CASE 1: Continue executing in a step/step-function after the first assertion has failed (by continuing to execute the remaining function)
* CASE 2: Continue executing additional steps in a scenario after the first step has failed

After reading your answer, you are looking for a solution of CASE 1 (which is the functionality that is provided/supported by `assertpy.soft_assertions`).
I understood that you want to solve the problem of CASE 2 (which is supported by the `Scenario.continue_after_failed_step` mechanism).

**Reply by jenisys** [reply]
@JSorella 
You stated above that the **ChatGPT solution did not work for you**.
I have not tried an example with `behave` step-functions, like above.
I explored a simpler example to use `soft_assertions` as function-decorator instead of a context-manager (using with-statement).

```python
from assertpy import assert_that, soft_assertions

@soft_assertions()   # IMPORTANT: Trailing parenthesis on the decorator (make the difference)
def example4soft_assertions_as_decorator(name):
    assert_that(name).is_equal_to("Alice")
    assert_that(name).is_equal_to("Bob")

def main():
    example4soft_assertions_as_decorator("Charly")
```

**But Python decorators are a bit tricky:** 
`@my_decorator` and `@my_decorator()` are two different things 
and only work in a similar way if the decorator implementation supports it.

If you do not add the parenthesis to the `@soft_assertions` as decorator,
the example-function (as function) is passed to the `soft_assertions()` call (which will be unexpected/not-supported).
The same problem should exist, if you use it with a step-function (like in the ChatGPT example above).

SEE ALSO:
* Python docs: [ContextManager used as decorator](https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator) (ContextDecorator examples)

**Reply by JSorella** [reply]
Hi @jenisys , 
I really appreciate your responses. I have tried everything you have explained.
But it seems that this soft-asserts solution is not compatible with Behave.
When this soft-assert fails, the rest of the test execution stops. So, many steps are not executed.

Is this a Behave limitation?

PD: What I am trying to achieve is the following:
- If a "soft assert" fails in a certain step, mark that step as failed, but continue the execution with the following steps.
- If a "hard assert" fails in a certain step, mark that steps as failed and stop the test execution (this is the default behavior in Python).
This is the "TestNG way" and I am trying to achieve this since the very first day I decided to work with Behave, but I am not currently able to find a solution for this.

**Comment by jenisys** [comment]
@JSorella 
A proof-of-concept example is provided under [behave/examples/soft_asserts](https://github.com/behave/behave/tree/main/examples/soft_asserts).
It shows all cases:

* CASE 1: Support soft-assertions within a step-function and continue after a failed assertion.
* CASE 2: Continue in Scenario after a step has failed.
* CASE 1 and 2: Combines both cases (probably what interests you)

SEE ALSO:

* [behave/examples/soft_asserts](https://github.com/behave/behave/tree/main/examples/soft_asserts)
* [behave_run.output_example.txt](https://github.com/behave/behave/tree/main/examples/soft_asserts/behave_run.output_example.txt)

HINTS:
* Works with Python3
* Works with Python2.7 if you use the ContextManager approach instead of the decorator approach.
* Probably, the Python 2.7 `contextlib` module seems to have a problem, if you use the decorator solution.

```shell
# -- HINT: Python 2.7 SYNDROME if you use @soft_assertions() decorator
Exception TypeError: 'GeneratorContextManager' object is not callable
Traceback (most recent call last):
   ...
  File "features/steps/number_steps.py", line 26, in <module>
    @soft_assertions()
TypeError: 'GeneratorContextManager' object is not callable
```

**Comment by JSorella** [comment]
Hi @jenisys,

I really appreciate your answers, are really helpful to me.
But I am sure I am not being clear about what I am trying to achieve.
In my automation suite I have lot of scenarios with many steps. Some scenarios have more than 20 steps on it.
This is a more detailed example of what I need to achieve:

```
  Scenario: Example scenario

    Given the user have the required data
    When I login into the system
    Then I should be in the expected page (if not, stop test execution)
    When I click over the items button
    Then I should see that the items sidebar is displayed  (if not, stop test execution)
    When I click over the first item
    Then I should see that 1 or more results are shown  (if not, continue with the execution anyway)
    When I click over the second item
    Then I should see a popup displayed  (if not, continue with the execution anyway)
    When I click over the third item
    Then I should see "Third Item" as title (if not, stop test execution)
```

~According with the examples you have uploaded recently, it seems I am not able to "mix" those two kinds of assertions in one scenario.~
~Also, the `@behave.continue_after_failed_step` it's not working for me (and I don't know why).~

I have achieved the "continue_after_failed_step" in the following way:
`context.scenario.continue_after_failed_step = True`

I have tried the following in order to achieve a `soft_assertion` in the way I need it:

```
@step('I should see that {expected_results_number} or more results are shown')
def step_impl(context, expected_results_number):
    context.scenario.continue_after_failed_step = True
    actual_results_number = context.page.get_visible_results_number()

    try:
        assert_that(actual_results_number).is_greater_than_or_equal_to(expected_results_number)
    except AssertionError as e:
        raise e
    finally:
        context.scenario.continue_after_failed_step = False
```
...of course this didn't work because when I change the `continue_after_failed_step` flag to False, the iteration of the next step will not be executed.
I need to find a way to change this flag to True or False depending on whether the validation step needs to stop execution there or continue with the rest of the scenario.
I could reset the flag to "False" in each scenario with "hard assertions" on it but I am pretty sure that would be an overkill (and a code mess).

Sorry if I'm taking this to a level where the framework isn't supposed to behave, but this is what I really need to achieve (or at least my development team expects me to achieve).

PD:
I am currently using Python 3.8.10

**Comment by JSorella** [comment]
Hi again @jenisys ,

I was working with the examples you have updated in the Behave repository to understand more deeply about the Behave behavior and what is capable to do.
I have understood that the `@behave.continue_after_failed_step` works only if you add a `if` sentence in the `before_scenario` (I was expecting not to do that to achieve this behavior, but at last, it worked fine for me).

The main issue with the `Scenario: Failing with Soft Assertions -- CASE 1 and CASE 2` example you have provided is that the Hard Assert (I mean, the `assert_that()` without the `soft_assertions()` decorator) is not causing the test execution to stop there: the execution will continue to the end of the scenario.

So, I have concluded that there are two possible ways using Behave:

1. Any failing assert, no matter the type of assert of their decorators, will cause the test execution to stop there.
2. Using the  `continue_after_failed_step` flag, the scenario will be executed to the bottom, no matter if any assert fails (including Hard Asserts).

I am trying to achieve an intermediate option, in which the Hard Asserts should cause the test execution to stop there (and not continue with further steps into the scenario) and the Soft Asserts should allow the test execution to continue with further steps (but marking the step as failed).
I think this "3rd" would be possible if, inside the `behave/model.py` file, the following line.... :

```
                        run_steps = (self.continue_after_failed_step and
                                     step.status == Status.failed)
```

... is modified as follows:
```
                        run_steps = ((self.continue_after_failed_step or step.continue_after_failed_step) and
                                     step.status == Status.failed)
```
.... But I don't know if its possible to add attributes for a certain step inside the step itself.
